### PR TITLE
Add search to integrations 🔍

### PR DESCRIPTION
**Pull request by timmo001**
## Proposed change

As requested :smile: 

integrations-search

## Type of change

- [x] New feature (thank you!)

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [x] Tests have been added to verify that the new code works.

### CODE REVIEW

File: src/panels/config/integrations/ha-config-integrations.ts [hunk 1]
```diff
@@ -360,7 +435,8 @@ class HaConfigIntegrations extends SubscribeMixin(LitElement) {
                       </ha-card>
                     `;
               })
-            : html`
+            : !this._configEntries
```
Comment (bramkragten) [inline src/panels/config/integrations/ha-config-integrations.ts:438]: Config entries will always be defined, it is inited as an empty array.

Comment (bramkragten) [inline src/panels/config/integrations/ha-config-integrations.ts:438]:
```suggestion
            : !this._configEntries.length
```

File: src/panels/config/integrations/ha-config-integrations.ts [hunk 2]
```diff
@@ -378,7 +454,28 @@ class HaConfigIntegrations extends SubscribeMixin(LitElement) {
                     >
                   </div>
                 </ha-card>
-              `}
+              `
+            : ""}
+          ${this._filter &&
+          !configEntriesInProgress.length &&
+          !configEntries.length
```
Comment (bramkragten) [inline src/panels/config/integrations/ha-config-integrations.ts:461]: && this._configEntries.length

File: src/panels/config/integrations/ha-config-integrations.ts [hunk 3]
```diff
@@ -100,6 +113,51 @@ class HaConfigIntegrations extends SubscribeMixin(LitElement) {
...
+        let keys: any[] = ["handler"];
+        configEntriesInProgress.map(
+          (entry: DataEntryFlowProgress): DataEntryFlowProgress => {
+            keys = mergeArrays(
+              keys,
+              Object.keys(entry.context).map((key: string) => `context.${key}`)
+            );
+            return entry;
+          }
+        );
```
Comment (bramkragten) [inline src/panels/config/integrations/ha-config-integrations.ts:147]: Let's restrict this to `localizeConfigFlowTitle(this.hass.localize, entry)` and `domainToName(this.hass.localize, entry.handler)` and maybe `entry.handler`.

**Comment by bramkragten** [comment]
Can we make the search bar look the same as the other config pages?

File: src/panels/config/integrations/ha-config-integrations.ts [hunk 4]
```diff
@@ -100,6 +113,51 @@ class HaConfigIntegrations extends SubscribeMixin(LitElement) {
     ];
   }
 
+  private _filterConfigEntries = memoizeOne(
+    (configEntries: ConfigEntry[], filter?: string): ConfigEntry[] => {
+      if (filter) {
+        const options: Fuse.FuseOptions<ConfigEntry> = {
+          keys: ["entry_id", "domain", "title", "source"],
```
[the remark this answers is not in the document]
Comment (timmo001) [reply src/panels/config/integrations/ha-config-integrations.ts]: I'll translate this in here then and swap out the HTML to the domain returned

**Comment by timmo001** [comment]
> Can we make the search bar look the same as the other config pages?

I can add `no-underline` but I don't think it suits this page:

Screenshot_20200423_135953

**Comment by timmo001** [comment]
Ah crud, broke those lovely images :smile:

**Comment by bramkragten** [comment]
I mean the search bar white, and in mobile in the header:
image

File: src/panels/config/integrations/ha-config-integrations.ts [hunk 5]
```diff
@@ -100,6 +112,69 @@ class HaConfigIntegrations extends SubscribeMixin(LitElement) {
...
   }
 
+  private _filterConfigEntries = memoizeOne(
+    (configEntries: ConfigEntry[], filter?: string): ConfigEntry[] => {
+      if (filter) {
+        let configEntriesSearch: ConfigEntry[] = configEntries.map(
+          (configEntry: ConfigEntry) => {
+            return {
+              ...configEntry,
+              domain: domainToName(this.hass.localize, configEntry.domain),
```
Comment (bramkragten) [inline src/panels/config/integrations/ha-config-integrations.ts:122]: Let's add this when we fetch the config entries and also use it in the render.

File: src/panels/config/integrations/ha-config-integrations.ts [hunk 6]
```diff
@@ -100,6 +112,69 @@ class HaConfigIntegrations extends SubscribeMixin(LitElement) {
...
+          threshold: 0.2,
+        };
+        const fuse = new Fuse(configEntriesSearch, options);
+        configEntriesSearch = fuse.search(filter);
+        configEntries = configEntries.filter(
+          (configEntry: ConfigEntry) =>
+            configEntriesSearch.findIndex(
+              (ce: ConfigEntry) => ce.entry_id === configEntry.entry_id
+            ) !== -1
+        );
```
Comment (bramkragten) [inline src/panels/config/integrations/ha-config-integrations.ts:139]: Also means we don't have to do this.

File: src/panels/config/integrations/ha-config-integrations.ts [hunk 7]
```diff
@@ -100,6 +112,69 @@ class HaConfigIntegrations extends SubscribeMixin(LitElement) {
...
+    }
+  );
+
+  private _filterConfigEntriesInProgress = memoizeOne(
+    (
+      configEntriesInProgress: DataEntryFlowProgress[],
+      filter?: string
+    ): DataEntryFlowProgress[] => {
+      if (filter) {
+        let configEntriedInProgressSearch: DataEntryFlowProgressSearch[] = configEntriesInProgress.map(
```
Comment (bramkragten) [inline src/panels/config/integrations/ha-config-integrations.ts:151]: Same here, add it in the subscribe function. Make a new extended type.

File: src/panels/config/integrations/ha-config-integrations.ts [hunk 8]
```diff
@@ -623,6 +764,18 @@ class HaConfigIntegrations extends SubscribeMixin(LitElement) {
...
+          left: -8px;
+          top: -7px;
+          display: flex;
+          align-items: center;
+          color: var(--secondary-text-color);
+          padding: 0 16px;
+        }
+        .search {
+          padding: 2px 16px;
+        }
```
Comment (bramkragten) [inline src/panels/config/integrations/ha-config-integrations.ts:778]:
```css
    background: var(--sidebar-background-color);
    border-bottom: 1px solid var(--divider-color);
```

File: src/panels/config/integrations/ha-config-integrations.ts (hunk 7 above)
Comment (timmo001) [reply src/panels/config/integrations/ha-config-integrations.ts]: Struggling here to get translations to load after subscribe change level (line 107). Its an async function is called but even if I try to move around and await that call (this.hass.loadBackendTranslation) it still returns blank, like the translation hasn't loaded yet.

The only way I've been able to get the translation is to load it at render or filter level like it is now.. :confused: 

Any ideas?

**Comment by timmo001** [comment]
Got it in a working state. Added suggestions all but `_configEntriesInProgress`. Having issue with loading titles in the subscribe function. (see above review comment)

**Comment by timmo001** [comment]
Ah promises. Did attempt those but I think I was having one of those days :smile: 

Thanks @bramkragten! :beers:

**Pull request merged**